akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestAppender.scala

```diff
@@ -0,0 +1,148 @@
+
+  def removeFilter(loggerName: String, filter: LoggingEventFilter): Unit =
+    getTestAppender(loggerName).removeTestFilter(filter)
+
+  private def loggerNameOrRoot(loggerName: String): String =
+    if (loggerName == "") org.slf4j.Logger.ROOT_LOGGER_NAME else loggerName
+
+  private def getLogbackLogger(loggerName: String): ch.qos.logback.classic.Logger = {
+    LoggerFactory.getLogger(loggerNameOrRoot(loggerName)) match {
+      case logger: ch.qos.logback.classic.Logger => logger
+      case null =>
+        throw new IllegalArgumentException(s"TestAppender couldn't find logger for [$loggerName].")
+      case other =>
+        throw new IllegalArgumentException(
+          s"TestAppender requires Logback logger for [$loggerName], " +
+          s"it was a [${other.getClass.getName}]")
+    }
+  }
+
+  private def getTestAppender(loggerName: String): TestAppender = {
+    val logger = getLogbackLogger(loggerName)
+    logger.getAppender(TestAppenderName) match {
+      case testAppender: TestAppender => testAppender
+      case null =>
+        throw new IllegalStateException(s"No $TestAppenderName was setup for logger [${logger.getName}]")
+      case other =>
+        throw new IllegalStateException(
+          s"Unexpected $TestAppenderName already added for logger [${logger.getName}]: $other")
+    }
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class TestAppender extends AppenderBase[ILoggingEvent] {
+
+  private var filters: List[LoggingEventFilter] = Nil
+
+  // invocations are synchronized via doAppend in AppenderBase
+  override def append(event: ILoggingEvent): Unit = {
+    import akka.util.ccompat.JavaConverters._
+
+    val throwable = event.getThrowableProxy match {
+      case p: ThrowableProxy =>
+        Option(p.getThrowable)
+      case _ => None
+    }
+
+    val loggingEvent = LoggingEvent(
+      level = convertLevel(event.getLevel),
+      message = event.getFormattedMessage,
+      loggerName = event.getLoggerName,
+      threadName = event.getThreadName,
+      timeStamp = event.getTimeStamp,
+      marker = Option(event.getMarker),
+      throwable = throwable,
+      mdc = event.getMDCPropertyMap.asScala.toMap)
+
+    filter(loggingEvent)
+  }
+
+  private def convertLevel(level: ch.qos.logback.classic.Level): Level = {
+    level.levelInt match {
+      case ch.qos.logback.classic.Level.TRACE_INT => Level.TRACE
+      case ch.qos.logback.classic.Level.DEBUG_INT => Level.DEBUG
+      case ch.qos.logback.classic.Level.INFO_INT  => Level.INFO
+      case ch.qos.logback.classic.Level.WARN_INT  => Level.WARN
+      case ch.qos.logback.classic.Level.ERROR_INT => Level.ERROR
+      case _ =>
+        throw new IllegalArgumentException("Level " + level.levelStr + ", " + level.levelInt + " is unknown.")
+    }
+  }
+
+  private def filter(event: LoggingEvent): Boolean = {
+    filters.exists(f =>
+      try {
+        f.apply(event)
+      } catch {
+        case _: Exception => false
+      })
+  }
```

Comment:
Mustn't this be synchronized as well to guarantee changes are visible?

Reply:
it's called from `append`, which synchronized via `doAppend` as noted in comment there
it's private, so I only made the public methods synchronized

Reply:
Alright, thanks for clarifying.